redisson-tomcat/redisson-tomcat-8/src/main/java/org/redisson/tomcat/RedissonSession.java

```diff
@@ -373,9 +373,10 @@ public void save() {
         if (broadcastSessionEvents) {
             newMap.put(IS_EXPIRATION_LOCKED, isExpirationLocked);
         }
-        
-        if (attrs != null) {
-            for (Entry<String, Object> entry : attrs.entrySet()) {
+
+        Map<String,Object> attrMap = readMode == ReadMode.REDIS ? loadedAttributes : attrs;
+        if (attrMap != null) {
+            for (Entry<String, Object> entry : attrMap.entrySet()) {
```

Comment:
So the question here is whether or not accessed objects (getAttribute) should beat written objects (setAttribute).

Let's say attribute X equals A.  Let's say you have two requests, Request 1 that sets attribute X and Request 2 that gets attribute X.  If Request 2 gets the attribute (returns A) and then Request 1 sets the attribute (sets to B), but Request 2 saves last, what should the attribute be set to?

I think it should be set to B, but in this case, it will be set to A.

Reply:
so the tricky part with that is that now actual object references are being returned via getAttribute calls, i can make modifications to them without ever calling setAttribute
```
Map<String, String> myMap = session.getAttribute("myStringMap");
myMap.put("new","value");
```
at the end of the request, should the map on the session, saved in redis, contain `new:value`?
I think we have to treat all getAttribute calls as potentially dirty